### PR TITLE
[Perf] Improve SyntaxRewriter visitation performance

### PR DESCRIPTION
**Pull request by rintaro**
* Introduce 'SyntaxNodeFactory'. This is an improved replacement of "`Syntax.Info` reusing mechanism" in `SyntaxVisitor`. Adopt it in both `SyntaxVisitor` and `SyntaxRewriter`
* Improve new layout creation:
  * Use a manually allocated 'UnsafeBufferPointer' to avoid 'Array' related overhead
  * Pre-initialize the buffer with old layout at once, then only update each rewritten child element. This is simpler and faster than appending each element.

### CODE REVIEW

**Comment by rintaro** [comment]
@swift-ci Please test

File: CodeGeneration/Sources/generate-swift-syntax/templates/swiftsyntax/SyntaxRewriterFile.swift [hunk 1]
```diff
@@ -329,10 +353,18 @@ let syntaxRewriterFile = SourceFileSyntax(leadingTrivia: copyrightHeader) {
           }
 
           // Build the Syntax node to rewrite
-          let absoluteRaw = AbsoluteRawSyntax(raw: child, info: info)
+          var childNode: Syntax
+          if !recyclableNodeInfos.isEmpty {
+            let recycledInfo: Syntax.Info = recyclableNodeInfos.removeLast()!
+            recycledInfo.info = .nonRoot(.init(parent: Syntax(node), absoluteInfo: info))
```
Comment (rintaro) [inline CodeGeneration/Sources/generate-swift-syntax/templates/swiftsyntax/SyntaxRewriterFile.swift:359]:
```suggestion
            recycledInfo.info = .nonRoot(.init(parent: syntaxNode, absoluteInfo: info))
```

File: CodeGeneration/Sources/generate-swift-syntax/templates/swiftsyntax/SyntaxRewriterFile.swift [hunk 2]
```diff
@@ -187,21 +216,16 @@ let syntaxRewriterFile = SourceFileSyntax(leadingTrivia: copyrightHeader) {
...
-          visitPost(node)
-        }
-        if let newNode = visitAny(node) {
-          return newNode
-        }
-        return Syntax(visit(castedNode))
+      ) {
+        let origNode = node
+        visitPre(origNode)
+        node = visitAny(origNode) ?? Syntax(visit(origNode.cast(NodeType.self)))
```
Comment (rintaro) [inline CodeGeneration/Sources/generate-swift-syntax/templates/swiftsyntax/SyntaxRewriterFile.swift:227]:
```suggestion
        node = visitAny(origNode) ?? visit(origNode.cast(NodeType.self))._syntaxNode
```

Comment (rintaro) [reply CodeGeneration/Sources/generate-swift-syntax/templates/swiftsyntax/SyntaxRewriterFile.swift:227]: No, no effect.

**Comment by rintaro** [comment]
Too many changes in this PR. I will split this into several PRs.

File: CodeGeneration/Sources/generate-swift-syntax/templates/swiftsyntax/SyntaxRewriterFile.swift [hunk 3]
```diff
@@ -42,6 +42,13 @@ let syntaxRewriterFile = SourceFileSyntax(leadingTrivia: copyrightHeader) {
       """
     )
 
+    DeclSyntax(
+      """
+      /// 'Syntax' object factory recycling 'Syntax.Info' instances.
+      private let nodeFactory: SyntaxNodeFactory = SyntaxNodeFactory()
```
Comment (rintaro) [inline CodeGeneration/Sources/generate-swift-syntax/templates/swiftsyntax/SyntaxRewriterFile.swift:48]: `SyntaxVisitor` should also use this as well.

**Comment by rintaro** [comment]
@swift-ci Please test

File: CodeGeneration/Sources/generate-swift-syntax/templates/swiftsyntax/SyntaxRewriterFile.swift [hunk 4]
```diff
@@ -305,77 +312,52 @@ let syntaxRewriterFile = SourceFileSyntax(leadingTrivia: copyrightHeader) {
...
+        for case let (child?, info) in RawSyntaxChildren(node) where viewMode.shouldTraverse(node: child) {
 
           // Build the Syntax node to rewrite
-          let absoluteRaw = AbsoluteRawSyntax(raw: child, info: info)
+          var childNode = nodeFactory.create(parent: node, raw: child, absoluteInfo: info)
 
-          let rewritten = dispatchVisit(Syntax(absoluteRaw, parent: syntaxNode))
-          if rewritten.id != info.nodeId {
+          dispatchVisit(&childNode)
+          if childNode.raw.id != child.id {
```
Comment (rintaro) [inline CodeGeneration/Sources/generate-swift-syntax/templates/swiftsyntax/SyntaxRewriterFile.swift:327]: Previously it compared `SyntaxIdentifier` which takes the entire tree into account.
Now it compares `RawSyntax.ID` because we only care the `RawSyntax` here.

**Comment by rintaro** [comment]
@swift-ci Please test

**Comment by rintaro** [comment]
@swift-ci Please test Windows

**Comment by rintaro** [comment]
@swift-ci Please test Windows

**Comment by rintaro** [comment]
@swift-ci Please clean test Windows

**Comment by rintaro** [comment]
@swift-ci Please test

**Comment by rintaro** [comment]
@swift-ci Please test

**Comment by rintaro** [comment]
@swift-ci Please test Linux

File: CodeGeneration/Sources/generate-swift-syntax/templates/swiftsyntax/SyntaxRewriterFile.swift [hunk 5]
```diff
@@ -187,21 +201,16 @@ let syntaxRewriterFile = SourceFileSyntax(leadingTrivia: copyrightHeader) {
       """
       /// Interpret `node` as a node of type `nodeType`, visit it, calling
       /// the `visit` to transform the node.
+      @inline(__always)
       private func visitImpl<NodeType: SyntaxProtocol>(
-        _ node: Syntax,
+        _ node: inout Syntax,
```
Comment (ahoppen) [inline CodeGeneration/Sources/generate-swift-syntax/templates/swiftsyntax/SyntaxRewriterFile.swift:206]: Out of curiosity: Do you know why `inout` is faster here? Does it avoid reference counting?

File: Sources/SwiftSyntax/generated/SyntaxRewriter.swift [hunk 6]
```diff
@@ -3898,79 +3909,52 @@ open class SyntaxRewriter {
...
     // nodes are being collected.
-    var newLayout: ContiguousArray<RawSyntax?>?
+    var newLayout: UnsafeMutableBufferPointer<RawSyntax?> = .init(start: nil, count: 0)
 
-    // Rewritten children just to keep their 'SyntaxArena' alive until they are
-    // wrapped with 'Syntax'
-    var rewrittens: ContiguousArray<Syntax> = []
+    // Keep 'SyntaxArena' of rewritten nodes alive until they are wrapped
+    // with 'Syntax'
+    var rewrittens: ContiguousArray<RetainedSyntaxArena> = []
```
Comment (ahoppen) [inline Sources/SwiftSyntax/generated/SyntaxRewriter.swift:3916]: What are the performance characteristics of `ContiguousArray` vs `Array` here? I would expect `Array` to be faster because it doesn’t need to move all elements if it runs out of space. And I don’t think we need the contiguous property here anymore.

File: Sources/SwiftSyntax/generated/SyntaxRewriter.swift [hunk 7]
```diff
@@ -3898,79 +3909,52 @@ open class SyntaxRewriter {
...
+      newLayout.deallocate()
       // 'withExtendedLifetime' to keep 'SyntaxArena's of them alive until here.
       return withExtendedLifetime(rewrittens) {
-        Syntax(raw: newRaw, rawNodeArena: arena).cast(SyntaxType.self)
+        Syntax(raw: newRaw, rawNodeArena: arena)
       }
     } else {
       // No child node was rewritten. So no need to change this node as well.
-      return node
+      return nil
```
Comment (ahoppen) [inline Sources/SwiftSyntax/generated/SyntaxRewriter.swift:3957]: Is there a benefit from returning `nil` vs `node` here? AFAICT we always to `?? node` on every call and we could avoid that if `visitChildren` didn’t return an optional.

File: Sources/SwiftSyntax/SyntaxArenaAllocatedBuffer.swift [hunk 8]
```diff
@@ -107,4 +107,8 @@ public struct SyntaxArenaAllocatedBufferPointer<Element: Sendable>: RandomAccess
   var unsafeRawBufferPointer: UnsafeRawBufferPointer {
     return UnsafeRawBufferPointer(buffer)
   }
+
+  public func withContiguousStorageIfAvailable<R>(_ body: (UnsafeBufferPointer<Element>) throws -> R) rethrows -> R? {
+    try body(buffer)
+  }
```
Comment (ahoppen) [inline Sources/SwiftSyntax/SyntaxArenaAllocatedBuffer.swift:113]: I this used anywhere?

File: cmake/modules/AddSwiftHostLibrary.cmake [hunk 9]
```diff
@@ -92,7 +92,7 @@ function(add_swift_syntax_library name)
   add_custom_command(
     TARGET ${target}
     POST_BUILD
-    COMMAND "${CMAKE_COMMAND}" -E touch_nocreate $<TARGET_FILE:${target}> $<TARGET_OBJECTS:${target}> "${module_base}"
```
Comment (ahoppen) [inline cmake/modules/AddSwiftHostLibrary.cmake:95]: Is this change related?

**Review by ahoppen — COMMENTED** [review]
Looks good. Just a couple minor comments.

File: cmake/modules/AddSwiftHostLibrary.cmake (hunk 9 above)
Comment (bnbarham) [reply cmake/modules/AddSwiftHostLibrary.cmake:95]: In an unfortunate way. The windows build fails with a fairly cryptic error, my best guess is that we're overflowing the command (either in number of arguments or length) with the extra file @rintaro added. This whole command is really just a hack to help with incremental builds, but we should check if it's needed in 3.26/3.29 and if not, I'm happy enough to just remove it.

**Comment by rintaro** [comment]
@swift-ci Please test Linux

File: CodeGeneration/Sources/generate-swift-syntax/templates/swiftsyntax/SyntaxRewriterFile.swift (hunk 5 above)
Comment (rintaro) [reply CodeGeneration/Sources/generate-swift-syntax/templates/swiftsyntax/SyntaxRewriterFile.swift:206]: I actually haven't measured performance difference between `inout` or not. I just followed `SyntaxVisitor` pattern here.

File: Sources/SwiftSyntax/generated/SyntaxRewriter.swift (hunk 6 above)
Comment (rintaro) [reply Sources/SwiftSyntax/generated/SyntaxRewriter.swift:3916]: Hm, I haven't tried `Array` here. interesting.

File: Sources/SwiftSyntax/generated/SyntaxRewriter.swift (hunk 7 above)
Comment (rintaro) [reply Sources/SwiftSyntax/generated/SyntaxRewriter.swift:3957]: Oh true, will try.

File: Sources/SwiftSyntax/SyntaxArenaAllocatedBuffer.swift (hunk 8 above)
Comment (rintaro) [reply Sources/SwiftSyntax/SyntaxArenaAllocatedBuffer.swift:113]: Yes, `UnsafeMutableBufferPointer.initialize(fromContentsOf:)` have a fast path calling this witness.

File: CodeGeneration/Sources/generate-swift-syntax/templates/swiftsyntax/SyntaxRewriterFile.swift (hunk 5 above)
Comment (rintaro) [reply CodeGeneration/Sources/generate-swift-syntax/templates/swiftsyntax/SyntaxRewriterFile.swift:206]: Also since we do `isKnownUniquelyReferenced`, we can't do like `let newNode = dispatchVisit(oldNode)` because `newNode` and `oldNode` might be the same. So we need: `node = dispatchVisit(node)` anyway. which is semantically what `inout` does.

File: Sources/SwiftSyntax/generated/SyntaxRewriter.swift (hunk 7 above)
Comment (rintaro) [reply Sources/SwiftSyntax/generated/SyntaxRewriter.swift:3957]: Ah e.g. `visitChildren(node._syntaxNode)?.cast(AccessorEffectSpecifiersSyntax.self) ?? node`
For this pattern, `.cast(AccessorEffectSpecifiersSyntax.self)` have some cost (even after `__shared` change). So being able to use `node` as-is was faster

File: CodeGeneration/Sources/generate-swift-syntax/templates/swiftsyntax/SyntaxRewriterFile.swift (hunk 5 above)
Comment (ahoppen) [reply CodeGeneration/Sources/generate-swift-syntax/templates/swiftsyntax/SyntaxRewriterFile.swift:206]: Ah, I see. Makes sense 👍🏽

File: Sources/SwiftSyntax/SyntaxArenaAllocatedBuffer.swift (hunk 8 above)
Comment (ahoppen) [reply Sources/SwiftSyntax/SyntaxArenaAllocatedBuffer.swift:113]: Ah, I didn’t think about this satisfying a defaulted requirement in `RandomAccessCollection`.

**Comment by rintaro** [comment]
@swift-ci Please test

**Comment by rintaro** [comment]
@swift-ci Please test

File: Sources/SwiftSyntax/generated/SyntaxRewriter.swift (hunk 6 above)
Comment (rintaro) [reply Sources/SwiftSyntax/generated/SyntaxRewriter.swift:3916]: Array always uses contiguous storage unless it's backed by `NSArray`.
Although there's no particular reason to use `ContiguousArray`, its implementation is simpler than `Array`. So let's keep this as-is

**Comment by rintaro** [comment]
@swift-ci Please test Windows

File: Sources/SwiftSyntax/generated/SyntaxRewriter.swift (hunk 7 above)
Comment (rintaro) [reply Sources/SwiftSyntax/generated/SyntaxRewriter.swift]: Sorry I re-measured and I don't see any difference returning `node` instead of `nil`. So I changed it to return `node` and remove `?? node` in the caller.

**Review by ahoppen — APPROVED** [review]
